File: .tmp_c4000/batch_04_go/3520_minimum_threshold_for_inversion_pairs_count.go
// LeetCode 3520 - Minimum Threshold for Inversion Pairs Count
// https://leetcode.com/problems/minimum-threshold-for-inversion-pairs-count/

func minThreshold(nums []int, k int) int {
	mx := 0
	for _, v := range nums {
		if v > mx {
			mx = v
		}
	}
	l, r := 0, mx+1
	for l < r {
		m := (l + r) / 2
		if countInv(nums, k, m) {
			r = m
		} else {
			l = m + 1
		}
	}
	if l > mx {
		return -1
	}
	return l
}
func countInv(nums []int, k, threshold int) bool {
	sorted := []int{}
	inv := 0
	for _, num := range nums {
		// count values in (num, num+threshold]
		lo, hi := 0, len(sorted)
		for lo < hi {
			mid := (lo + hi) / 2
			if sorted[mid] <= num {
				lo = mid + 1
			} else {
				hi = mid
			}
		}
		left := lo
		lo, hi = 0, len(sorted)
		for lo < hi {
			mid := (lo + hi) / 2
			if sorted[mid] <= num+threshold {
				lo = mid + 1
			} else {
				hi = mid
			}
		}
		inv += lo - left
		// insert num
		lo, hi = 0, len(sorted)
		for lo < hi {
			mid := (lo + hi) / 2
			if sorted[mid] < num {
				lo = mid + 1
			} else {
				hi = mid
			}
		}
		sorted = append(sorted, 0)
		copy(sorted[lo+1:], sorted[lo:])
		sorted[lo] = num
	}
	return inv >= k
}

File: .tmp_py4013/batch_13_dump/3520_minimum_threshold_for_inversion_pairs_count.js
// LeetCode 3520 - Minimum Threshold for Inversion Pairs Count
// https://leetcode.com/problems/minimum-threshold-for-inversion-pairs-count/

function upperBound(a, target) {
    let lo = 0, hi = a.length;
    while (lo < hi) {
        const mid = (lo + hi) >> 1;
        if (a[mid] <= target) lo = mid + 1;
        else hi = mid;
    }
    return lo;
}
function countInv(nums, k, threshold) {
    const sorted = [];
    let inv = 0;
    for (const num of nums) {
        const left = upperBound(sorted, num);
        const right = upperBound(sorted, num + threshold);
        inv += right - left;
        sorted.splice(upperBound(sorted, num), 0, num);
    }
    return inv >= k;
}
var minThreshold = function(nums, k) {
    let mx = 0;
    for (const v of nums) if (v > mx) mx = v;
    let l = 0, r = mx + 1;
    while (l < r) {
        const m = (l + r) >> 1;
        if (countInv(nums, k, m)) r = m;
        else l = m + 1;
    }
    return l > mx ? -1 : l;
};
